File: example/CoreGraphics.ts
export class CGPoint {
  constructor(readonly x: number, readonly y: number) {}
}

export class CGSize {
  constructor(readonly width: number, readonly height: number) {}
}

export class CGRect {
  constructor(
    args:
      | { origin: CGPoint; size: CGSize }
      | {
          x: number;
          y: number;
          width: number;
          height: number;
          origin?: undefined;
          size?: undefined;
        }
  ) {
    this.origin = args.origin ?? new CGPoint(args.x, args.y);
    this.size = args.size ?? new CGSize(args.width, args.height);
  }

  origin: CGPoint;
  size: CGSize;

  contains(point: CGPoint): boolean {
    return (
      this.origin.x <= point.x &&
      point.x <= this.origin.x + this.size.width &&
      this.origin.y <= point.y &&
      point.y <= this.origin.y + this.size.height
    );
  }
}

File: example/Sources/MyExample/CoreGraphics.swift
import Foundation
import NodeAPI

extension CGFloat: NodeValueConvertible, NodeValueCreatable {
  public typealias ValueType = NodeNumber
  
  public func nodeValue() throws -> any NodeAPI.NodeValue {
    try NodeNumber(Double(self))
  }
  
  public static func from(_ value: NodeAPI.NodeNumber) throws -> CGFloat {
    CGFloat(try value.double())
  }
}

extension CGPoint: NodeValueConvertible, NodeValueCreatable {
  public typealias ValueType = NodeObject
  
  public func nodeValue() throws -> any NodeValue {
    try NodeObject([
      "x": x,
      "y": y,
    ])
  }
  
  public static func from(_ value: ValueType) throws -> Self {
    let x = try value.propertyAs("x", Double.self)
    let y = try value.propertyAs("y", Double.self)
    return Self(x: x, y: y)
  }
}


extension CGSize: NodeValueConvertible, NodeValueCreatable {
  public typealias ValueType = NodeObject
  
  public func nodeValue() throws -> any NodeValue {
    try NodeObject([
      "width": width,
      "height": height,
    ])
  }
  
  public static func from(_ value: ValueType) throws -> Self {
    let width = try value.propertyAs("width", Double.self)
    let height = try value.propertyAs("height", Double.self)
    return Self(width: width, height: height)
  }
}

extension CGRect: NodeValueConvertible, NodeValueCreatable {
  public typealias ValueType = NodeObject
  
  public func nodeValue() throws -> any NodeValue {
    try NodeObject([
      "origin": origin,
      "size": size,
    ])
  }
  
  public static func from(_ value: ValueType) throws -> Self {
    let origin = try value.propertyAs("origin", CGPoint.self)
    let size = try value.propertyAs("size", CGSize.self)
    return Self(origin: origin, size: size)
  }
}
